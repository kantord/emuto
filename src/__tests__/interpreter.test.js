// @flow

import execute from '../interpreter'
import compile from '../compiler'
import parse from '../parsers/parser'

const tests = [
  {
    sourceCode: `null`,
    output: null
  },
  {
    sourceCode: `true`,
    output: true
  },
  {
    sourceCode: `$`,
    input: true,
    output: true
  },
  {
    sourceCode: `$`,
    input: 98,
    output: 98
  },
  {
    sourceCode: `.bar`,
    input: { bar: 4, baz: 'hello' },
    output: 4
  },
  {
    sourceCode: `.$shit`,
    input: { $shit: null, baz: 'hello' },
    output: null
  },
  {
    sourceCode: `[0.14, true, "true 2"]`,
    output: [0.14, true, 'true 2']
  },
  {
    sourceCode: `[[false], [true, null]]`,
    output: [[false], [true, null]]
  },
  {
    sourceCode: `false | true`,
    output: true
  },
  {
    sourceCode: `.foo | .bar`,
    input: {
      foo: {
        bar: 'hello'
      }
    },
    output: 'hello'
  },
  {
    sourceCode: `.foo | .bar | .baz`,
    input: {
      foo: {
        bar: {
          baz: 'world'
        }
      }
    },
    output: 'world'
  },
  {
    sourceCode: `.apple.pear`,
    input: {
      apple: {
        pear: 42
      }
    },
    output: 42
  },
  {
    sourceCode: `.apple.pear.bar`,
    input: {
      apple: {
        pear: { bar: false }
      }
    },
    output: false
  },
  {
    sourceCode: `null: true`,
    output: [null, true]
  },
  {
    sourceCode: `{
      "foo": [-42, $],
      "bar": {
        "message": "Hello World"
      }
    }`,
    input: 42,
    output: {
      foo: [-42, 42],
      bar: { message: 'Hello World' }
    }
  },
  {
    sourceCode: `{"foo": ("bar" | "baz")}`,
    output: {
      foo: 'baz'
    }
  },
  {
    sourceCode: `{"foo": ("bar" | "baz")}["foo"]`,
    output: 'baz'
  },
  {
    sourceCode: `[3, 2][0]`,
    output: 3
  },
  {
    sourceCode: `[3, 2][0] : [3, 2][1]`,
    output: [3, 2]
  },
  {
    sourceCode: `[3, 2][0] | [3, 2][1]`,
    output: 2
  },
  {
    sourceCode: `join " "`,
    input: ['Hello', 'World'],
    output: 'Hello World'
  },
  {
    sourceCode: `map \\" "`,
    input: ['Hello', 'World'],
    output: [' ', ' ']
  },
  {
    sourceCode: `{"original": $.foo} | {"new": $}`,
    input: { foo: 42 },
    output: { new: { original: 42 } }
  },
  {
    sourceCode: `map $ => {"original": $.foo} | map $=> {"new": $}`,
    input: [{ foo: 42 }, { foo: 'hello' }],
    output: [{ new: { original: 42 } }, { new: { original: 'hello' } }]
  },
  {
    sourceCode: `map $ => {"original": $.foo} | "foo"`,
    input: [{ foo: 42 }, { foo: 'hello' }],
    output: 'foo'
  },
  {
    sourceCode: `[0,1,2][1,2][1,1][0]`,
    output: 2
  },
  {
    sourceCode: `[0,1,"foo"][1,2][1,1][0]`,
    output: 'foo'
  },
  {
    sourceCode: `{"foo": [{"bar": "baz"}]} | $.foo[0].bar`,
    output: 'baz'
  },
  {
    sourceCode: `sortBy \\.age | map \\.name`,
    output: ['John', 'Mary'],
    input: [{ name: 'John', age: 11 }, { name: 'Mary', age: 33 }]
  },
  {
    sourceCode: `8 * 3.14 / 2`,
    output: 8 * 3.14 / 2
  },
  {
    sourceCode: `8 + 3.14 * 2`,
    output: 8 + 3.14 * 2
  },
  {
    sourceCode: `(8 + 3.14) * 2`,
    output: (8 + 3.14) * 2
  },
  {
    sourceCode: `8 - (3.14 * 2)`,
    output: 8 - 3.14 * 2
  },
  {
    sourceCode: `8314 % 34 -3`,
    output: 8314 % 34 - 3
  },
  {
    sourceCode: `3 * 2 -1 <= -1.34 * 3`,
    output: 3 * 2 - 1 <= -1.34 * 3
  },
  {
    sourceCode: `3 * 2 -1 < -1.34 * 3`,
    output: 3 * 2 - 1 < -1.34 * 3
  },
  {
    sourceCode: `3 * 2 -1 >= -1.34 * 3`,
    output: 3 * 2 - 1 >= -1.34 * 3
  },
  {
    sourceCode: `3 * 2 -1 > -1.34 * 3`,
    output: 3 * 2 - 1 > -1.34 * 3
  },
  {
    sourceCode: `3 * 2 -1 == -1.34 * 3`,
    output: 3 * 2 - 1 === -1.34 * 3
  },
  {
    sourceCode: `3 * 2 -1 != -1.34 * 3`,
    output: 3 * 2 - 1 !== -1.34 * 3
  },
  {
    sourceCode: `"1" == 1`,
    output: '1' === 1
  },
  {
    sourceCode: `3 * 2 -1 > -1.34 * 3 || true`,
    output: 3 * 2 - 1 > -1.34 * 3 || true
  },
  {
    sourceCode: `3 * 2 -1 > -1.34 * 3 || false`,
    output: 3 * 2 - 1 > -1.34 * 3 || false
  },
  {
    sourceCode: `3 * 2 -1 > -1.34 * 3 && false`,
    output: 3 * 2 - 1 > -1.34 * 3 && false
  },
  {
    sourceCode: `3 * 2 -1 > -1.34 * 3 && true`,
    output: 3 * 2 - 1 > -1.34 * 3 && true
  },
  {
    sourceCode: `-(3 * 2) -1 > -1.34 * 3 && true`,
    output: -(3 * 2) - 1 > -1.34 * 3 && true
  },
  {
    sourceCode: `+(3 * 2) -1 > -1.34 * 3 && true`,
    output: +(3 * 2) - 1 > -1.34 * 3 && true
  },
  {
    sourceCode: `!(3 * 2 -1 > -1.34 * 3) && true`,
    output: !(3 * 2 - 1 > -1.34 * 3) && true
  },
  {
    sourceCode: `!(3 * .foo -1 > -1.34 * .bar) && true`,
    input: {
      foo: 2,
      bar: 3
    },
    output: !(3 * 2 - 1 > -1.34 * 3) && true
  },
  {
    sourceCode: `filter $ => .age >= 18 | map $ => .name`,
    input: [
      { name: 'John', age: 17 },
      { name: 'Mary', age: 18 },
      { name: 'Gabe', age: 19 },
      { name: 'Franz', age: 8 }
    ],
    output: ['Mary', 'Gabe']
  },
  {
    sourceCode: `3 * 2 -1 > -1.34 * 3 or true`,
    output: 3 * 2 - 1 > -1.34 * 3 || true
  },
  {
    sourceCode: `!(3 * 2 -1 > -1.34 * 3) and true`,
    output: !(3 * 2 - 1 > -1.34 * 3) && true
  },
  {
    sourceCode: `3 + $foobar where $foobar = 4`,
    output: 7
  },
  {
    sourceCode: `(3 + $foobar | [$, $foobar]) where $foobar = $`,
    input: 4,
    output: [7, 4]
  },
  {
    sourceCode: `($x + $foobar | [$, $foobar]) where $x = 3 $foobar = ($x + 1)`,
    input: 4,
    output: [7, 4]
  },
  {
    sourceCode: `([1] | map $foo | $[0]) where $foo = ($ => 4)`,
    output: 4
  },
  {
    sourceCode: `([1] | foo 4) where $foo = ($ => $ => 4)`,
    output: 4
  },
  {
    sourceCode: `([1, 2] | foo) where $foo = $reverse`,
    output: [2, 1]
  },
  {
    sourceCode: `map $foo => [$foo]`,
    input: [1, 2],
    output: [[1], [2]]
  },
  {
    sourceCode: `reduce ($a => $b => $a + $b): 0`,
    input: [1, 2, -1, 3],
    output: 5
  },
  {
    sourceCode: `[1, ...$, 3, ...$]`,
    input: [1, 2],
    output: [1, 1, 2, 3, 1, 2]
  },
  {
    sourceCode: `{"foo": 3} | {"bar": "baz", ...$}`,
    output: {
      foo: 3,
      bar: 'baz'
    }
  },
  {
    sourceCode: `$.uno?.dos?.tres`,
    input: {
      uno: {}
    },
    output: null
  },
  {
    sourceCode: `$?[3]`,
    input: undefined,
    output: null
  },
  {
    sourceCode: `$?[3]`,
    input: [1, 2, 3],
    output: null
  }
]

describe('interpreter', () => {
  tests.forEach(
    ({
      sourceCode,
      output,
      input
    }: {
      sourceCode: string,
      output: mixed,
      input?: mixed
    }) => {
      it(`executes ${sourceCode}`, () => {
        expect(execute(sourceCode)(input)).toEqual(output)
      })

      it(`correct target code ${sourceCode}`, () => {
        expect(compile(sourceCode)).toMatchSnapshot()
      })

      it(`correct target tree ${sourceCode}`, () => {
        expect(parse(sourceCode)).toMatchSnapshot()
      })
    }
  )

  it('throws syntax error on invalid source code', () => {
    expect(() => {
      execute('ยก')
    }).toThrow(SyntaxError)
  })
})
