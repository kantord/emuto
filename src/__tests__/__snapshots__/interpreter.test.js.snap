// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interpreter correct target code $ 1`] = `"(function(_) { return (function(input) { return input})})"`;

exports[`interpreter correct target code $ 2`] = `"(function(_) { return (function(input) { return input})})"`;

exports[`interpreter correct target code .$shit 1`] = `"(function(_) { return (function(input) { return input.$shit})})"`;

exports[`interpreter correct target code .apple.pear 1`] = `"(function(_) { return (function(input) { return input.apple.pear})})"`;

exports[`interpreter correct target code .apple.pear.bar 1`] = `"(function(_) { return (function(input) { return input.apple.pear.bar})})"`;

exports[`interpreter correct target code .bar 1`] = `"(function(_) { return (function(input) { return input.bar})})"`;

exports[`interpreter correct target code .foo | .bar | .baz 1`] = `"(function(_) { return (function(input) { return (function (input) {return (function (input) {return input.baz})(input.bar)})(input.foo)})})"`;

exports[`interpreter correct target code .foo | .bar 1`] = `"(function(_) { return (function(input) { return (function (input) {return input.bar})(input.foo)})})"`;

exports[`interpreter correct target code [[false], [true, null]] 1`] = `"(function(_) { return (function(input) { return [[false], [true, null]]})})"`;

exports[`interpreter correct target code [0.14, true, "true 2"] 1`] = `"(function(_) { return (function(input) { return [0.14, true, \\"true 2\\"]})})"`;

exports[`interpreter correct target code [3, 2][0] : [3, 2][1] 1`] = `"(function(_) { return (function(input) { return [_.projection([3, 2], [0]),_.projection([3, 2], [1])]})})"`;

exports[`interpreter correct target code [3, 2][0] | [3, 2][1] 1`] = `"(function(_) { return (function(input) { return (function (input) {return _.projection([3, 2], [1])})(_.projection([3, 2], [0]))})})"`;

exports[`interpreter correct target code [3, 2][0] 1`] = `"(function(_) { return (function(input) { return _.projection([3, 2], [0])})})"`;

exports[`interpreter correct target code {
      "foo": [-42, $],
      "bar": {
        "message": "Hello World"
      }
    } 1`] = `"(function(_) { return (function(input) { return (_.objectify([[\\"foo\\",[-42, input]], [\\"bar\\",(_.objectify([[\\"message\\",\\"Hello World\\"]]))]]))})})"`;

exports[`interpreter correct target code {"foo": ("bar" | "baz")} 1`] = `"(function(_) { return (function(input) { return (_.objectify([[\\"foo\\",((function (input) {return \\"baz\\"})(\\"bar\\"))]]))})})"`;

exports[`interpreter correct target code {"foo": ("bar" | "baz")}["foo"] 1`] = `"(function(_) { return (function(input) { return _.projection((_.objectify([[\\"foo\\",((function (input) {return \\"baz\\"})(\\"bar\\"))]])), [\\"foo\\"])})})"`;

exports[`interpreter correct target code {"original": $.foo} | {"new": $} 1`] = `"(function(_) { return (function(input) { return (function (input) {return (_.objectify([[\\"new\\",input]]))})((_.objectify([[\\"original\\",input.foo]])))})})"`;

exports[`interpreter correct target code false | true 1`] = `"(function(_) { return (function(input) { return (function (input) {return true})(false)})})"`;

exports[`interpreter correct target code join " " 1`] = `"(function(_) { return (function(input) { return _.join(\\" \\")(input)})})"`;

exports[`interpreter correct target code map $ => {"original": $.foo} | "foo" 1`] = `"(function(_) { return (function(input) { return (function (input) {return \\"foo\\"})((_.map(function(input) {return (_.objectify([[\\"original\\",input.foo]]))})(input)))})})"`;

exports[`interpreter correct target code map $ => {"original": $.foo} | map $=> {"new": $} 1`] = `"(function(_) { return (function(input) { return (function (input) {return (_.map(function(input) {return (_.objectify([[\\"new\\",input]]))})(input))})((_.map(function(input) {return (_.objectify([[\\"original\\",input.foo]]))})(input)))})})"`;

exports[`interpreter correct target code map \\" " 1`] = `"(function(_) { return (function(input) { return (_.map(function(input) {return \\" \\"})(input))})})"`;

exports[`interpreter correct target code null 1`] = `"(function(_) { return (function(input) { return null})})"`;

exports[`interpreter correct target code null: true 1`] = `"(function(_) { return (function(input) { return [null,true]})})"`;

exports[`interpreter correct target code true 1`] = `"(function(_) { return (function(input) { return true})})"`;

exports[`interpreter correct target tree $ 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "input",
    "value": "$",
  },
}
`;

exports[`interpreter correct target tree $ 2`] = `
Object {
  "status": true,
  "value": Object {
    "name": "input",
    "value": "$",
  },
}
`;

exports[`interpreter correct target tree .$shit 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "inputProp",
    "value": ".$shit",
  },
}
`;

exports[`interpreter correct target tree .apple.pear 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "valueProp",
    "value": Object {
      "left": Object {
        "name": "inputProp",
        "value": ".apple",
      },
      "right": ".pear",
    },
  },
}
`;

exports[`interpreter correct target tree .apple.pear.bar 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "valueProp",
    "value": Object {
      "left": Object {
        "name": "inputProp",
        "value": ".apple",
      },
      "right": ".pear.bar",
    },
  },
}
`;

exports[`interpreter correct target tree .bar 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "inputProp",
    "value": ".bar",
  },
}
`;

exports[`interpreter correct target tree .foo | .bar | .baz 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "pipe",
    "value": Object {
      "left": Object {
        "name": "inputProp",
        "value": ".foo",
      },
      "right": Object {
        "name": "pipe",
        "value": Object {
          "left": Object {
            "name": "inputProp",
            "value": ".bar",
          },
          "right": Object {
            "name": "inputProp",
            "value": ".baz",
          },
        },
      },
    },
  },
}
`;

exports[`interpreter correct target tree .foo | .bar 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "pipe",
    "value": Object {
      "left": Object {
        "name": "inputProp",
        "value": ".foo",
      },
      "right": Object {
        "name": "inputProp",
        "value": ".bar",
      },
    },
  },
}
`;

exports[`interpreter correct target tree [[false], [true, null]] 1`] = `
Object {
  "status": true,
  "value": Object {
    "end": Object {
      "column": 24,
      "line": 1,
      "offset": 23,
    },
    "name": "list",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
    "value": Array [
      Object {
        "end": Object {
          "column": 9,
          "line": 1,
          "offset": 8,
        },
        "name": "list",
        "start": Object {
          "column": 2,
          "line": 1,
          "offset": 1,
        },
        "value": Array [
          Object {
            "name": "primitive",
            "value": "false",
          },
        ],
      },
      Object {
        "end": Object {
          "column": 23,
          "line": 1,
          "offset": 22,
        },
        "name": "list",
        "start": Object {
          "column": 11,
          "line": 1,
          "offset": 10,
        },
        "value": Array [
          Object {
            "name": "primitive",
            "value": "true",
          },
          Object {
            "name": "primitive",
            "value": "null",
          },
        ],
      },
    ],
  },
}
`;

exports[`interpreter correct target tree [0.14, true, "true 2"] 1`] = `
Object {
  "status": true,
  "value": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "name": "list",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
    "value": Array [
      Object {
        "name": "primitive",
        "value": "0.14",
      },
      Object {
        "name": "primitive",
        "value": "true",
      },
      Object {
        "name": "primitive",
        "value": "\\"true 2\\"",
      },
    ],
  },
}
`;

exports[`interpreter correct target tree [3, 2][0] : [3, 2][1] 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "tuple",
    "value": Array [
      Object {
        "name": "projection",
        "value": Object {
          "left": Object {
            "end": Object {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "name": "list",
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "3",
              },
              Object {
                "name": "primitive",
                "value": "2",
              },
            ],
          },
          "right": Object {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "name": "list",
            "start": Object {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "0",
              },
            ],
          },
        },
      },
      Object {
        "name": "projection",
        "value": Object {
          "left": Object {
            "end": Object {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "name": "list",
            "start": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "3",
              },
              Object {
                "name": "primitive",
                "value": "2",
              },
            ],
          },
          "right": Object {
            "end": Object {
              "column": 22,
              "line": 1,
              "offset": 21,
            },
            "name": "list",
            "start": Object {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "1",
              },
            ],
          },
        },
      },
    ],
  },
}
`;

exports[`interpreter correct target tree [3, 2][0] | [3, 2][1] 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "pipe",
    "value": Object {
      "left": Object {
        "name": "projection",
        "value": Object {
          "left": Object {
            "end": Object {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "name": "list",
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "3",
              },
              Object {
                "name": "primitive",
                "value": "2",
              },
            ],
          },
          "right": Object {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "name": "list",
            "start": Object {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "0",
              },
            ],
          },
        },
      },
      "right": Object {
        "name": "projection",
        "value": Object {
          "left": Object {
            "end": Object {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "name": "list",
            "start": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "3",
              },
              Object {
                "name": "primitive",
                "value": "2",
              },
            ],
          },
          "right": Object {
            "end": Object {
              "column": 22,
              "line": 1,
              "offset": 21,
            },
            "name": "list",
            "start": Object {
              "column": 19,
              "line": 1,
              "offset": 18,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "1",
              },
            ],
          },
        },
      },
    },
  },
}
`;

exports[`interpreter correct target tree [3, 2][0] 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "projection",
    "value": Object {
      "left": Object {
        "end": Object {
          "column": 7,
          "line": 1,
          "offset": 6,
        },
        "name": "list",
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
        "value": Array [
          Object {
            "name": "primitive",
            "value": "3",
          },
          Object {
            "name": "primitive",
            "value": "2",
          },
        ],
      },
      "right": Object {
        "end": Object {
          "column": 10,
          "line": 1,
          "offset": 9,
        },
        "name": "list",
        "start": Object {
          "column": 7,
          "line": 1,
          "offset": 6,
        },
        "value": Array [
          Object {
            "name": "primitive",
            "value": "0",
          },
        ],
      },
    },
  },
}
`;

exports[`interpreter correct target tree {
      "foo": [-42, $],
      "bar": {
        "message": "Hello World"
      }
    } 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "object",
    "value": Array [
      Object {
        "name": "tuple",
        "value": Array [
          Object {
            "name": "primitive",
            "value": "\\"foo\\"",
          },
          Object {
            "end": Object {
              "column": 22,
              "line": 2,
              "offset": 23,
            },
            "name": "list",
            "start": Object {
              "column": 14,
              "line": 2,
              "offset": 15,
            },
            "value": Array [
              Object {
                "name": "primitive",
                "value": "-42",
              },
              Object {
                "name": "input",
                "value": "$",
              },
            ],
          },
        ],
      },
      Object {
        "name": "tuple",
        "value": Array [
          Object {
            "name": "primitive",
            "value": "\\"bar\\"",
          },
          Object {
            "name": "object",
            "value": Array [
              Object {
                "name": "tuple",
                "value": Array [
                  Object {
                    "name": "primitive",
                    "value": "\\"message\\"",
                  },
                  Object {
                    "name": "primitive",
                    "value": "\\"Hello World\\"",
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
}
`;

exports[`interpreter correct target tree {"foo": ("bar" | "baz")} 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "object",
    "value": Array [
      Object {
        "name": "tuple",
        "value": Array [
          Object {
            "name": "primitive",
            "value": "\\"foo\\"",
          },
          Object {
            "name": "parentheses",
            "value": Object {
              "name": "pipe",
              "value": Object {
                "left": Object {
                  "name": "primitive",
                  "value": "\\"bar\\"",
                },
                "right": Object {
                  "name": "primitive",
                  "value": "\\"baz\\"",
                },
              },
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`interpreter correct target tree {"foo": ("bar" | "baz")}["foo"] 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "projection",
    "value": Object {
      "left": Object {
        "name": "object",
        "value": Array [
          Object {
            "name": "tuple",
            "value": Array [
              Object {
                "name": "primitive",
                "value": "\\"foo\\"",
              },
              Object {
                "name": "parentheses",
                "value": Object {
                  "name": "pipe",
                  "value": Object {
                    "left": Object {
                      "name": "primitive",
                      "value": "\\"bar\\"",
                    },
                    "right": Object {
                      "name": "primitive",
                      "value": "\\"baz\\"",
                    },
                  },
                },
              },
            ],
          },
        ],
      },
      "right": Object {
        "end": Object {
          "column": 32,
          "line": 1,
          "offset": 31,
        },
        "name": "list",
        "start": Object {
          "column": 25,
          "line": 1,
          "offset": 24,
        },
        "value": Array [
          Object {
            "name": "primitive",
            "value": "\\"foo\\"",
          },
        ],
      },
    },
  },
}
`;

exports[`interpreter correct target tree {"original": $.foo} | {"new": $} 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "pipe",
    "value": Object {
      "left": Object {
        "name": "object",
        "value": Array [
          Object {
            "name": "tuple",
            "value": Array [
              Object {
                "name": "primitive",
                "value": "\\"original\\"",
              },
              Object {
                "name": "valueProp",
                "value": Object {
                  "left": Object {
                    "name": "input",
                    "value": "$",
                  },
                  "right": ".foo",
                },
              },
            ],
          },
        ],
      },
      "right": Object {
        "name": "object",
        "value": Array [
          Object {
            "name": "tuple",
            "value": Array [
              Object {
                "name": "primitive",
                "value": "\\"new\\"",
              },
              Object {
                "name": "input",
                "value": "$",
              },
            ],
          },
        ],
      },
    },
  },
}
`;

exports[`interpreter correct target tree false | true 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "pipe",
    "value": Object {
      "left": Object {
        "name": "primitive",
        "value": "false",
      },
      "right": Object {
        "name": "primitive",
        "value": "true",
      },
    },
  },
}
`;

exports[`interpreter correct target tree join " " 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "functionCall",
    "value": Object {
      "left": Object {
        "name": "identifier",
        "value": "join",
      },
      "right": Object {
        "name": "primitive",
        "value": "\\" \\"",
      },
    },
  },
}
`;

exports[`interpreter correct target tree map $ => {"original": $.foo} | "foo" 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "pipe",
    "value": Object {
      "left": Object {
        "name": "functionCallLambda",
        "value": Object {
          "left": Object {
            "name": "identifier",
            "value": "map",
          },
          "right": Object {
            "name": "object",
            "value": Array [
              Object {
                "name": "tuple",
                "value": Array [
                  Object {
                    "name": "primitive",
                    "value": "\\"original\\"",
                  },
                  Object {
                    "name": "valueProp",
                    "value": Object {
                      "left": Object {
                        "name": "input",
                        "value": "$",
                      },
                      "right": ".foo",
                    },
                  },
                ],
              },
            ],
          },
        },
      },
      "right": Object {
        "name": "primitive",
        "value": "\\"foo\\"",
      },
    },
  },
}
`;

exports[`interpreter correct target tree map $ => {"original": $.foo} | map $=> {"new": $} 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "pipe",
    "value": Object {
      "left": Object {
        "name": "functionCallLambda",
        "value": Object {
          "left": Object {
            "name": "identifier",
            "value": "map",
          },
          "right": Object {
            "name": "object",
            "value": Array [
              Object {
                "name": "tuple",
                "value": Array [
                  Object {
                    "name": "primitive",
                    "value": "\\"original\\"",
                  },
                  Object {
                    "name": "valueProp",
                    "value": Object {
                      "left": Object {
                        "name": "input",
                        "value": "$",
                      },
                      "right": ".foo",
                    },
                  },
                ],
              },
            ],
          },
        },
      },
      "right": Object {
        "name": "functionCallLambda",
        "value": Object {
          "left": Object {
            "name": "identifier",
            "value": "map",
          },
          "right": Object {
            "name": "object",
            "value": Array [
              Object {
                "name": "tuple",
                "value": Array [
                  Object {
                    "name": "primitive",
                    "value": "\\"new\\"",
                  },
                  Object {
                    "name": "input",
                    "value": "$",
                  },
                ],
              },
            ],
          },
        },
      },
    },
  },
}
`;

exports[`interpreter correct target tree map \\" " 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "functionCallLambda",
    "value": Object {
      "left": Object {
        "name": "identifier",
        "value": "map",
      },
      "right": Object {
        "name": "primitive",
        "value": "\\" \\"",
      },
    },
  },
}
`;

exports[`interpreter correct target tree null 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "primitive",
    "value": "null",
  },
}
`;

exports[`interpreter correct target tree null: true 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "tuple",
    "value": Array [
      Object {
        "name": "primitive",
        "value": "null",
      },
      Object {
        "name": "primitive",
        "value": "true",
      },
    ],
  },
}
`;

exports[`interpreter correct target tree true 1`] = `
Object {
  "status": true,
  "value": Object {
    "name": "primitive",
    "value": "true",
  },
}
`;
